src/include/txn_inline.h

```diff
@@ -724,9 +724,8 @@ __txn_visible_id(WT_SESSION_IMPL *session, uint64_t id)
     if (txn->isolation == WT_ISO_READ_UNCOMMITTED)
         return (true);
 
-    /* Otherwise, we need a snapshot; if we don't have one, assume a test for global visibility. */
-    if (!F_ISSET(txn, WT_TXN_HAS_SNAPSHOT))
-        return (__txn_visible_all_id(session, id));
+    /* Otherwise, we should be called with a snapshot. */
+    WT_ASSERT(session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) || session->dhandle->checkpoint != NULL);
```

Comment:
Do you know why this path checks for a checkpoint handle?

Reply:
No, I don't. Let me remove this check and see how it goes.

Reply:
The reason I ask -- all of the plans for getting checkpoint cursors back involve the checkpoint cursor having an appropriate snapshot. :)

Reply:
It hits:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007f9f856068b1 in __GI_abort () at abort.c:79
#2  0x00007f9f843c796c in __wt_abort (session=0x5556a4ccf930) at ../src/os_common/os_abort.c:30
#3  0x00007f9f8424dd46 in __txn_visible_id (session=0x5556a4ccf930, id=5) at ../src/include/txn_inline.h:728
#4  0x00007f9f8424ddeb in __wt_txn_visible (session=0x5556a4ccf930, id=5, timestamp=50) at ../src/include/txn_inline.h:747
#5  0x00007f9f8424dba9 in __wt_txn_tw_start_visible (session=0x5556a4ccf930, tw=0x7ffe390f9da0)
    at ../src/include/txn_inline.h:637
#6  0x00007f9f8424e67d in __wt_txn_read (session=0x5556a4ccf930, cbt=0x5556a5143150, key=0x0, recno=1, upd=0x0)
    at ../src/include/txn_inline.h:1054
#7  0x00007f9f842526d1 in __cursor_var_next (cbt=0x5556a5143150, newpage=true, restart=false, skippedp=0x7ffe390f9f10)
    at ../src/btree/bt_curnext.c:312
#8  0x00007f9f84253ed1 in __wt_btcur_next_prefix (cbt=0x5556a5143150, prefix=0x0, truncating=false)
    at ../src/btree/bt_curnext.c:775
#9  0x00007f9f8425460e in __wt_btcur_next (cbt=0x5556a5143150, truncating=false) at ../src/btree/bt_curnext.c:893
#10 0x00007f9f84320121 in __curfile_next (cursor=0x5556a5143150) at ../src/cursor/cur_file.c:93
#11 0x00007f9f8473633f in _wrap_Cursor_next ()
```
I don't understand why we still can use the transaction if we don't have one.

Reply:
Not terribly surprising.  Thank you for checking!

Reply:
The failure does come from the checkpoint cursor.

Reply:
The existing checkpoint cursor code is slightly weird in this regard; effectively it acts like it has a snapshot (that comes from the checkpoint) but there isn't actually one there.